### ISSUE TITLE
A C++20 version of tight_pair

### ISSUE THREAD

**Issue by Morwenn** [Morwenn/tight_pair#4]
C++20 brings enough improvements to ship a C++20 version of `tight_pair` different from the C++17 one:
* `[[no_unique_address]]` makes the whole EBO implementation optional. It's worth noting that EBO has more guarantees about actually reducing the size of the class, but still. Having two implementations in `cruft::cxx_2017` and `cruft::cxx_2020` namespaces (with one `inline` depending on the version of the language) seems reasonable enough.
* `explicit(bool)` will definitely make the implementation of conditionally `explicit` constructors easier.
* We will definitely have to support `operator<=>` for the library to behave smoothly with everything else. Maybe it can even be used to simplify the implementation.
* Since ADL will be able to find function templates more easily, `using cruft::get` might not be needed anymore in the test suite (it's worth having a dedicated test).
* With more optional `typename` everywhere, the implementation can get less verbose.
* `std::endian` means that we can drop our hand-rolled implementation in the C++20 implementation.
* `std::remove_cvref` can surely be used here and there to slightly improve the compiler throughput.
* The new `<version>` header would be the right one to include for some feature-test macros (including non-standard ones like `_LIBCPP_VERSION`).
* Concepts can be used to properly constrain some constructors such as those accepting pair-like or tuple-like objects (we can build on the *tuple-like* exposition-only concept described in [P1035][1]).
* `requires` expressions can surely replace some of the remaining `std::enable_if_t` all over the place.
* [LWG2899](https://wg21.link/LWG2899) tweaks `noexcept` rules of some standard library wrapper types, check whether it affects us.

It's worth nothing that C++20 does bring a few improvements to `std::pair` already (*e.g.* more `constexpr`) so the README will also have to be edited to mention those: we don't want to belittle the recent improvements to `std::pair` :)

  [1]: wg21.link/P1035

**Comment by FirstLoveLife** [comment]
> `[[no_unique_address]]`

Yes, this is awesome feature. But does `compressed_pair` still be useful when `[[no_unique_address]]` is available? As you said, the size of class is not guaranteed to reduce, but IMO, that's major what it for, and, compilers will implement it, just as ebco. 

> `explicit(bool)`

I love this it.

> We will definitely have to support operator<=> for the library to behave smoothly with everything else. Maybe it can even be used to simplify the implementation.

I use https://github.com/taocpp/operators as an c++11/14/17 alternative:
```
template<typename T1, typename T2>
	struct tight_pair
		: private detail::tight_pair_storage<T1, T2>,
		  private Math::AbstractAlgebra::RelationAlgebra::totally_ordered<
			  tight_pair<T1, T2>>
```

which is also a c++11 version of boost/operator. 

> Since ADL will be able to find function templates more easily, using cruft::get might not be needed anymore in the test suite (it's worth having a dedicated test).

Does c++2a have feature about adl? I'll appreciate if you give some detailed explanations :)


> With more optional typename everywhere, the implementation can get less verbose.

Here is some my small utilities: 
```
	template<typename T>
	using _t = typename T::type;

	template<typename T>
	using _p = typename T::pointer;

	template<typename T>
	using _e = typename T::element_type;

	template<typename T>
	using _p = typename T::pointer;



#define IS(name) \
	template<typename T> \
	inline constexpr bool is_##name##_v = is_##name<T>::value;

#define BI_IS(name) \
	template<typename From, typename To> \
	inline constexpr bool is_##name##_v = is_##name<From, To>::value;

#define NOT(name) \
	template<typename T> \
	inline constexpr bool not_##name##_v = not is_##name<T>::value; \
	template<typename T> \
	struct not_##name : bool_<not_##name##_v<T>> \
	{};

#define BI_NOT(name) \
	template<typename From, typename To> \
	inline constexpr bool not_##name##_v = not is_##name<From, To>::value; \
	template<typename From, typename To> \
	struct not_##name : bool_<not_##name##_v<From, To>> \
	{};

#define ARE(name) \
	template<typename... T> \
	inline constexpr bool are_##name##_v = (is_##name##_v<T> && ...); \
\
	template<typename... T> \
	struct are_##name : bool_<are_##name##_v<T...>> \
	{};

#define BI_ARE(name) \
	template<typename... T> \
	inline constexpr bool are_##name##_v \
		= detail::binaryTraitAre_impl<is_##name, T...>(); \
\
	template<typename... T> \
	struct are_##name : bool_<are_##name##_v<T...>> \
	{};

#define ANY(name) \
	template<typename... T> \
	inline constexpr bool any_##name##_v = (is_##name##_v<T> || ...); \
\
	template<typename... T> \
	struct any_##name : bool_<any_##name##_v<T...>> \
	{};

#define BI_ANY(name) \
	template<typename... T> \
	inline constexpr bool any_##name##_v \
		= detail::binaryTraitOr_impl<is_##name, T...>(); \
\
	template<typename... T> \
	struct any_##name : bool_<any_##name##_v<T...>> \
	{};


#define IS_NOT_ARE_ANY(name) \
	IS(name) \
	NOT(name) \
	ARE(name) \
	ANY(name)

#define NOT_ARE_ANY(name) \
	NOT(name) \
	ARE(name) \
	ANY(name)

#define BI_IS_NOT_ARE_ANY(name) \
	BI_IS(name) \
	BI_NOT(name) \
	BI_ARE(name) \
	BI_ANY(name)

```

, whihc make it easy to define trait helper variable/type. 


> std::endian means that we can drop our hand-rolled implementation in the C++20 implementation.

I noted that your implementation is the same as proposal of `std::endian`'s. But I cannot understand your `struct can_optimize_compare`. What's behind little/big/native here?

**Comment by Morwenn** [comment]
To address the first point: once `[[no_unique_address]]` is out the whole `tight_pair` will probably be less valuable, but it wasn't meant to be used in real-world projects anyway, it's at the top of the README:

> *It is not meant to be anywhere near serious, but it was a fun little project to try to design a C++17-only component, and to optimize it as possible despite the underlying layers of abstraction :)*

The reason why I didn't use any external libraries (save for the test suite) is more or less the same: it's just among the arbitrary constraints I picked when starting this toy project (same reason I don't use macro to generate code while it would be the most pragmatic way to do things sometimes). To be honest I only maintain it so that I can have fun playing with the new C++ features, hence this whole issue to start with :p

Concerning ADL there is a change in C++20 due to [P0846](https://wg21.link/P0846) being adopted. The difference is that now function templates can be found by ADL even when a template parameter is manually specified. If I'm not mistaken, the following code doesn't work in C++17 but does in C++20:

```cpp
#include <tight_pair.h>

int main()
{
    cruft::tight_pair<int, int> foo;
    get<0>(foo);  // ADL can now find this call without a using
}
```

The `can_optimize_compare` mechanism is a dirty trick to "optimize" unsigned integer comparisons (even though benchmarks showed no real improvement, but it was still fun): when you've got a pair of unsigned integers and want to perform lexicographical comparisons (what pair does) on them, you can simply reinterpret their bits as an unsigned integer twice as big and compare the two bigger unsigned integers to actually perform the lexicographical comparison. But in order to reinterpret those safely, the two unsigned integers need to be ordered such that the bit reinterpretation works, which depends on endianness, hence the following conditions for the optimization to happen:
* The pair must store two unsigned integers of the same type
* There needs to exists an unsigned integer twice as big for which the comparison is efficient
* The unsigned integer shall not have padding bits
* The order of the unsigned integers must respect the endianness so that the most significant integer for lexicographical ordering appears in the position of the most significant unsigned integer bits once it's been reinterpreted to a bigger unsigned integer
